src/arch/riscv/sbi.c

```diff
@@ -230,9 +230,12 @@ struct sbiret sbi_time_handler(unsigned long fid)
     if (fid != SBI_SET_TIMER_FID) return (struct sbiret){SBI_ERR_NOT_SUPPORTED};
 
     uint64_t stime_value = vcpu_readreg(cpu()->vcpu, REG_A0);
-
-    sbi_set_timer(stime_value);  // assumes always success
-    CSRC(CSR_HVIP, HIP_VSTIP);
+    if(cpu_arch_has_extension(SSTC_EXTENSION)) {
+        CSRW(CSR_VSTIMECMP, stime_value);
+    } else {
+        sbi_set_timer(stime_value);  // assumes always success
+        CSRC(CSR_HVIP, HIP_VSTIP);
+    }
     CSRS(sie, SIE_STIE);
```

Comment:
Shouldn't this be inside the else branch above? We don't need to enable the hs-level timer interrupt if sstc is available...

Reply:
Fixed